solution-segmentation.py

```diff
@@ -0,0 +1,329 @@
+
+class Rectangle:
+    """Represents a geometric rectangle"""
+
+    def __init__(self, top_left, bottom_right):
+        """Constructs a rectangle.
+
+        :param top_left: A tuple representing the top left point in the rectangle.
+        :param bottom_right: A tuple representing the bottom right point in the rectangle.
+        """
+        self._top_left = top_left
+        self._bottom_right = bottom_right
+
+    @property
+    def tl(self):
+        """The top left point of the rectangle"""
+        return self._top_left
+
+    @property
+    def br(self):
+        """The bottom right point of the rectangle"""
+        return self._bottom_right
+
+    def x_slice(self):
+        """Extract a slice object for the x-range of the rectangle"""
+        return slice(self.tl[0], self.br[0])
+
+    def y_slice(self):
+        """Extract a slice object for the y-range of the rectangle"""
+        return slice(self.tl[1], self.br[1])
+
+
+def get_sampling_rectangle(img_shape, rect_size=(80, 100)):
+    """Computes the sampling rectangle based on the image and rectangle sizes
+
+    :param img_shape: The shape of the images, as returned by numpy.ndarray.shape.
+    :param rect_size: The size of the sampling rectangle given as the tuple (height, width)
+
+    :return A Rectangle representing the sampling rectangle
+    """
+
+    img_height, img_width, _ = img_shape
+    rect_height, rect_width = rect_size
+
+    center_x = img_width // 2
+    center_y = (img_height * 4) // 5
+    x_left = np.clip(center_x - rect_width // 2, 0, img_width)
+    x_right = np.clip(x_left + rect_width, 0, img_width)
+    y_top = np.clip(center_y - rect_height // 2, 0, img_height)
+    y_bottom = np.clip(y_top + rect_height, 0, img_height)
+
+    return Rectangle((x_left, y_top), (x_right, y_bottom))
+
+
+def extract_features(feature_image):
+    """Extracts features from the image frame
+
+    :param feature_image: The original image frame
+
+    :return An image of feature vectors in the np.float32 datatype
+    """
+
+    # Convert to float32.
+    feature_image = np.float32(feature_image) / 255.0
+
+    # Choose a colour format:
+    # return feature_image
+    # return cv2.cvtColor(feature_image, cv2.COLOR_BGR2HSV)
+    # return cv2.cvtColor(feature_image, cv2.COLOR_BGR2HLS)
+    # return cv2.cvtColor(feature_image, cv2.COLOR_BGR2Lab)
+    # return cv2.cvtColor(feature_image, cv2.COLOR_BGR2Luv)
+    # return cv2.cvtColor(feature_image, cv2.COLOR_BGR2XYZ)
+    return cv2.cvtColor(feature_image, cv2.COLOR_BGR2YCrCb)
+
+
+def extract_training_samples(feature_image, sampling_rectangle):
+    """Extracts training samples from a sampling rectangle
+
+    :param feature_image: An image of feature vectors.
+    :param sampling_rectangle: The region in the feature image to extract samples from.
+
+    :return The samples
+    """
+
+    patch = feature_image[sampling_rectangle.y_slice(), sampling_rectangle.x_slice()]
+    samples = patch.reshape(-1, 3)
+    return samples
```

Comment:
Spør for å lære: Er det riktig at man forstatt jobber på et "view" her, altså ikke tar noen kopi?

Reply:
Ja, jeg tror både `patch` og `samples` skal være "views" her